Firestore/core/src/firebase/firestore/remote/stream.mm

```diff
@@ -0,0 +1,318 @@
+               kBackoffInitialDelay, kBackoffMaxDelay},
+      idle_timer_id_{idle_timer_id} {
+}
+
+// Starting
+
+void Stream::Start() {
+  EnsureOnQueue();
+
+  if (state_ == State::Error) {
+    BackoffAndTryRestarting();
+    return;
+  }
+
+  LOG_DEBUG("%s start", GetDebugDescription());
+
+  HARD_ASSERT(state_ == State::Initial, "Already started");
+  state_ = State::Starting;
+
+  Authenticate();
+}
+
+void Stream::Authenticate() {
+  EnsureOnQueue();
+
+  // Auth may outlive the stream, so make sure it doesn't try to access a
+  // deleted object.
+  std::weak_ptr<Stream> weak_this{shared_from_this()};
+  int auth_generation = generation_;
+  credentials_provider_->GetToken(
+      [weak_this, auth_generation](StatusOr<Token> maybe_token) {
+        auto strong_this = weak_this.lock();
+        if (!strong_this) {
+          return;
+        }
+        strong_this->worker_queue_->EnqueueRelaxed(
+            [maybe_token, weak_this, auth_generation] {
+              auto strong_this = weak_this.lock();
+              // Streams can be stopped while waiting for authorization, so need
+              // to check generation.
+              if (!strong_this || strong_this->generation_ != auth_generation) {
+                return;
+              }
+              strong_this->ResumeStartAfterAuth(maybe_token);
+            });
+      });
+}
+
+void Stream::ResumeStartAfterAuth(const StatusOr<Token>& maybe_token) {
+  EnsureOnQueue();
+
+  HARD_ASSERT(state_ == State::Starting,
+              "State should still be 'Starting' (was %s)", state_);
+
+  if (!maybe_token.ok()) {
+    OnStreamError(maybe_token.status());
+    return;
+  }
+
+  absl::string_view token = [&] {
+    auto token = maybe_token.ValueOrDie();
+    return token.user().is_authenticated() ? token.token()
+                                           : absl::string_view{};
+  }();
+
+  grpc_stream_ = CreateGrpcStream(datastore_, token);
+  grpc_stream_->Start();
+}
+
+void Stream::OnStreamStart() {
+  EnsureOnQueue();
+
+  state_ = State::Open;
+  DoOnStreamStart();
+}
+
+// Backoff
+
+void Stream::BackoffAndTryRestarting() {
+  EnsureOnQueue();
+
+  LOG_DEBUG("%s backoff", GetDebugDescription());
+
+  HARD_ASSERT(state_ == State::Error,
+              "Should only perform backoff in an error case");
+
+  state_ = State::Backoff;
+  backoff_.BackoffAndRun([this] { ResumeStartFromBackoff(); });
+}
+
+void Stream::ResumeStartFromBackoff() {
+  EnsureOnQueue();
+
+  HARD_ASSERT(state_ == State::Backoff, "Backoff elapsed but state is now: %s",
+              state_);
+
+  state_ = State::Initial;
+  Start();
+  HARD_ASSERT(IsStarted(), "Stream should have started.");
+}
+
+void Stream::InhibitBackoff() {
+  EnsureOnQueue();
+
+  HARD_ASSERT(!IsStarted(),
+              "Can only cancel backoff in a stopped state (was %s)", state_);
+
+  // Clear the error condition.
+  state_ = State::Initial;
+  ResetBackoff();
+}
+
+void Stream::ResetBackoff() {
```

Comment:
After some thinking, I decided to revert the change. Basically, while I still think it's a (very) small improvement, I can also see your points, and I'm no longer convinced it's worthwhile. It's a very valid point that having `Inhibit/ResetBackoff` is confusing. Worse, while `MarkHealthy` seems to me a good way to describe what that function is doing, I'm not sure the reverse is true -- I would intuitively expect a function called `MarkHealthy` to do more than just adjust the backoff. I slightly disagree about the cognitive overload, because in my view exposing a data member to derived classes adds all of its public methods to the interface of the class. On the other hand, it's true that these classes are already tightly coupled by design. All in all, I guess I'll just leave things as they are now.